File: tests/test_builder_control_flow.py
#!/usr/bin/env -S uv run
"""
Test: test_builder_control_flow
Tests LLVM Builder control flow instructions

Python equivalent of tests/test_builder_control_flow.cpp
Must produce identical output to the C++ version.
"""

import llvm


def main():
    with llvm.create_context() as ctx:
        with ctx.create_module("test_builder_control_flow") as mod:
            i32 = ctx.types.i32
            i1 = ctx.types.i1
            void_ty = ctx.types.void

            with ctx.create_builder() as builder:
                # ==========================================
                # Function 1: void return
                # ==========================================
                void_func_ty = ctx.types.function(void_ty, [])
                void_func = mod.add_function("return_void", void_func_ty)
                void_entry = void_func.append_basic_block("entry", ctx)
                builder.position_at_end(void_entry)
                builder.ret_void()

                # ==========================================
                # Function 2: value return
                # ==========================================
                ret_func_ty = ctx.types.function(i32, [i32])
                ret_func = mod.add_function("return_value", ret_func_ty)
                ret_param = ret_func.get_param(0)
                ret_param.name = "x"

                ret_entry = ret_func.append_basic_block("entry", ctx)
                builder.position_at_end(ret_entry)
                builder.ret(ret_param)

                # ==========================================
                # Function 3: unconditional branch
                # ==========================================
                br_func = mod.add_function("unconditional_branch", void_func_ty)
                br_entry = br_func.append_basic_block("entry", ctx)
                br_target = br_func.append_basic_block("target", ctx)

                builder.position_at_end(br_entry)
                br_inst = builder.br(br_target)

                builder.position_at_end(br_target)
                builder.ret_void()

                # ==========================================
                # Function 4: conditional branch
                # ==========================================
                cond_func_ty = ctx.types.function(i32, [i1])
                cond_func = mod.add_function("conditional_branch", cond_func_ty)
                cond_param = cond_func.get_param(0)
                cond_param.name = "cond"

                cond_entry = cond_func.append_basic_block("entry", ctx)
                cond_true = cond_func.append_basic_block("if_true", ctx)
                cond_false = cond_func.append_basic_block("if_false", ctx)

                builder.position_at_end(cond_entry)
                cond_br = builder.cond_br(cond_param, cond_true, cond_false)

                builder.position_at_end(cond_true)
                builder.ret(i32.constant(1))

                builder.position_at_end(cond_false)
                builder.ret(i32.constant(0))

                # ==========================================
                # Function 5: switch statement
                # ==========================================
                switch_func_ty = ctx.types.function(i32, [i32])
                switch_func = mod.add_function("switch_example", switch_func_ty)
                switch_param = switch_func.get_param(0)
                switch_param.name = "val"

                switch_entry = switch_func.append_basic_block("entry", ctx)
                case_0 = switch_func.append_basic_block("case_0", ctx)
                case_1 = switch_func.append_basic_block("case_1", ctx)
                case_2 = switch_func.append_basic_block("case_2", ctx)
                default_case = switch_func.append_basic_block("default", ctx)

                builder.position_at_end(switch_entry)
                switch_inst = builder.switch_(switch_param, default_case, 3)
                switch_inst.add_case(i32.constant(0), case_0)
                switch_inst.add_case(i32.constant(1), case_1)
                switch_inst.add_case(i32.constant(2), case_2)

                builder.position_at_end(case_0)
                builder.ret(i32.constant(100))

                builder.position_at_end(case_1)
                builder.ret(i32.constant(200))

                builder.position_at_end(case_2)
                builder.ret(i32.constant(300))

                builder.position_at_end(default_case)
                builder.ret(i32.constant(-1, sign_extend=True))

                # ==========================================
                # Function 6: function call
                # ==========================================
                call_func = mod.add_function("call_example", ret_func_ty)
                call_param = call_func.get_param(0)
                call_param.name = "n"

                call_entry = call_func.append_basic_block("entry", ctx)
                builder.position_at_end(call_entry)

                args = [call_param]
                call_result = builder.call(ret_func_ty, ret_func, args, "result")
                builder.ret(call_result)

                # ==========================================
                # Function 7: unreachable
                # ==========================================
                unreach_func = mod.add_function("unreachable_example", void_func_ty)
                unreach_entry = unreach_func.append_basic_block("entry", ctx)
                builder.position_at_end(unreach_entry)
                builder.unreachable()

                # Check insert block
                current_block = builder.insert_block
                assert current_block is not None

            # Verify module
            if not mod.verify():
                print(f"; Verification failed: {mod.get_verification_error()}")
                return 1

            # Print diagnostic comments
            print("; Test: test_builder_control_flow")
            print(";")
            print("; Control flow operations demonstrated:")
            print(";   ret void, ret value")
            print(";   br (unconditional)")
            print(";   br (conditional)")
            print(";   switch with 3 cases + default")
            print(";   call")
            print(";   unreachable")
            print(";")
            print("; Branch analysis:")
            print(
                f";   unconditional br is conditional: {'yes' if br_inst.is_conditional else 'no'}"
            )
            print(
                f";   conditional br is conditional: {'yes' if cond_br.is_conditional else 'no'}"
            )
            print(f";   unconditional br num successors: {br_inst.num_successors}")
            print(f";   conditional br num successors: {cond_br.num_successors}")
            print(";")
            print(f"; Current insert block: {current_block.name}")
            print()

            # Print module IR
            print(mod.to_string(), end="")

    return 0


if __name__ == "__main__":
    exit(main())

File: tests/test_builder_control_flow.cpp
/**
 * Test: test_builder_control_flow
 * Tests LLVM Builder control flow instructions
 *
 * LLVM-C APIs covered:
 * - LLVMBuildRetVoid(), LLVMBuildRet()
 * - LLVMBuildBr(), LLVMBuildCondBr()
 * - LLVMBuildSwitch(), LLVMAddCase()
 * - LLVMBuildCall2()
 * - LLVMBuildUnreachable()
 * - LLVMGetInsertBlock()
 * - LLVMIsConditional(), LLVMGetCondition()
 * - LLVMGetNumSuccessors(), LLVMGetSuccessor()
 */

#include <cstdio>
#include <llvm-c/Analysis.h>
#include <llvm-c/Core.h>

int main() {
  LLVMContextRef ctx = LLVMContextCreate();
  LLVMModuleRef mod =
      LLVMModuleCreateWithNameInContext("test_builder_control_flow", ctx);

  LLVMTypeRef i32 = LLVMInt32TypeInContext(ctx);
  LLVMTypeRef i1 = LLVMInt1TypeInContext(ctx);
  LLVMTypeRef void_ty = LLVMVoidTypeInContext(ctx);

  LLVMBuilderRef builder = LLVMCreateBuilderInContext(ctx);

  // ==========================================
  // Function 1: void return
  // ==========================================
  LLVMTypeRef void_func_ty = LLVMFunctionType(void_ty, nullptr, 0, 0);
  LLVMValueRef void_func = LLVMAddFunction(mod, "return_void", void_func_ty);
  LLVMBasicBlockRef void_entry =
      LLVMAppendBasicBlockInContext(ctx, void_func, "entry");
  LLVMPositionBuilderAtEnd(builder, void_entry);
  LLVMBuildRetVoid(builder);

  // ==========================================
  // Function 2: value return
  // ==========================================
  LLVMTypeRef ret_params[] = {i32};
  LLVMTypeRef ret_func_ty = LLVMFunctionType(i32, ret_params, 1, 0);
  LLVMValueRef ret_func = LLVMAddFunction(mod, "return_value", ret_func_ty);
  LLVMValueRef ret_param = LLVMGetParam(ret_func, 0);
  LLVMSetValueName2(ret_param, "x", 1);

  LLVMBasicBlockRef ret_entry =
      LLVMAppendBasicBlockInContext(ctx, ret_func, "entry");
  LLVMPositionBuilderAtEnd(builder, ret_entry);
  LLVMBuildRet(builder, ret_param);

  // ==========================================
  // Function 3: unconditional branch
  // ==========================================
  LLVMValueRef br_func =
      LLVMAddFunction(mod, "unconditional_branch", void_func_ty);
  LLVMBasicBlockRef br_entry =
      LLVMAppendBasicBlockInContext(ctx, br_func, "entry");
  LLVMBasicBlockRef br_target =
      LLVMAppendBasicBlockInContext(ctx, br_func, "target");

  LLVMPositionBuilderAtEnd(builder, br_entry);
  LLVMValueRef br_inst = LLVMBuildBr(builder, br_target);

  LLVMPositionBuilderAtEnd(builder, br_target);
  LLVMBuildRetVoid(builder);

  // ==========================================
  // Function 4: conditional branch
  // ==========================================
  LLVMTypeRef cond_params[] = {i1};
  LLVMTypeRef cond_func_ty = LLVMFunctionType(i32, cond_params, 1, 0);
  LLVMValueRef cond_func =
      LLVMAddFunction(mod, "conditional_branch", cond_func_ty);
  LLVMValueRef cond_param = LLVMGetParam(cond_func, 0);
  LLVMSetValueName2(cond_param, "cond", 4);

  LLVMBasicBlockRef cond_entry =
      LLVMAppendBasicBlockInContext(ctx, cond_func, "entry");
  LLVMBasicBlockRef cond_true =
      LLVMAppendBasicBlockInContext(ctx, cond_func, "if_true");
  LLVMBasicBlockRef cond_false =
      LLVMAppendBasicBlockInContext(ctx, cond_func, "if_false");

  LLVMPositionBuilderAtEnd(builder, cond_entry);
  LLVMValueRef cond_br =
      LLVMBuildCondBr(builder, cond_param, cond_true, cond_false);

  LLVMPositionBuilderAtEnd(builder, cond_true);
  LLVMBuildRet(builder, LLVMConstInt(i32, 1, 0));

  LLVMPositionBuilderAtEnd(builder, cond_false);
  LLVMBuildRet(builder, LLVMConstInt(i32, 0, 0));

  // ==========================================
  // Function 5: switch statement
  // ==========================================
  LLVMTypeRef switch_params[] = {i32};
  LLVMTypeRef switch_func_ty = LLVMFunctionType(i32, switch_params, 1, 0);
  LLVMValueRef switch_func =
      LLVMAddFunction(mod, "switch_example", switch_func_ty);
  LLVMValueRef switch_param = LLVMGetParam(switch_func, 0);
  LLVMSetValueName2(switch_param, "val", 3);

  LLVMBasicBlockRef switch_entry =
      LLVMAppendBasicBlockInContext(ctx, switch_func, "entry");
  LLVMBasicBlockRef case_0 =
      LLVMAppendBasicBlockInContext(ctx, switch_func, "case_0");
  LLVMBasicBlockRef case_1 =
      LLVMAppendBasicBlockInContext(ctx, switch_func, "case_1");
  LLVMBasicBlockRef case_2 =
      LLVMAppendBasicBlockInContext(ctx, switch_func, "case_2");
  LLVMBasicBlockRef default_case =
      LLVMAppendBasicBlockInContext(ctx, switch_func, "default");

  LLVMPositionBuilderAtEnd(builder, switch_entry);
  LLVMValueRef switch_inst =
      LLVMBuildSwitch(builder, switch_param, default_case, 3);
  LLVMAddCase(switch_inst, LLVMConstInt(i32, 0, 0), case_0);
  LLVMAddCase(switch_inst, LLVMConstInt(i32, 1, 0), case_1);
  LLVMAddCase(switch_inst, LLVMConstInt(i32, 2, 0), case_2);

  LLVMPositionBuilderAtEnd(builder, case_0);
  LLVMBuildRet(builder, LLVMConstInt(i32, 100, 0));

  LLVMPositionBuilderAtEnd(builder, case_1);
  LLVMBuildRet(builder, LLVMConstInt(i32, 200, 0));

  LLVMPositionBuilderAtEnd(builder, case_2);
  LLVMBuildRet(builder, LLVMConstInt(i32, 300, 0));

  LLVMPositionBuilderAtEnd(builder, default_case);
  LLVMBuildRet(builder, LLVMConstInt(i32, -1, 1));

  // ==========================================
  // Function 6: function call
  // ==========================================
  LLVMValueRef call_func = LLVMAddFunction(mod, "call_example", ret_func_ty);
  LLVMValueRef call_param = LLVMGetParam(call_func, 0);
  LLVMSetValueName2(call_param, "n", 1);

  LLVMBasicBlockRef call_entry =
      LLVMAppendBasicBlockInContext(ctx, call_func, "entry");
  LLVMPositionBuilderAtEnd(builder, call_entry);

  LLVMValueRef args[] = {call_param};
  LLVMValueRef call_result =
      LLVMBuildCall2(builder, ret_func_ty, ret_func, args, 1, "result");
  LLVMBuildRet(builder, call_result);

  // ==========================================
  // Function 7: unreachable
  // ==========================================
  LLVMValueRef unreach_func =
      LLVMAddFunction(mod, "unreachable_example", void_func_ty);
  LLVMBasicBlockRef unreach_entry =
      LLVMAppendBasicBlockInContext(ctx, unreach_func, "entry");
  LLVMPositionBuilderAtEnd(builder, unreach_entry);
  LLVMBuildUnreachable(builder);

  // Check insert block
  LLVMBasicBlockRef current_block = LLVMGetInsertBlock(builder);

  LLVMDisposeBuilder(builder);

  // Verify module
  char *error = nullptr;
  if (LLVMVerifyModule(mod, LLVMReturnStatusAction, &error)) {
    fprintf(stderr, "; Verification failed: %s\n", error);
    LLVMDisposeMessage(error);
    LLVMDisposeModule(mod);
    LLVMContextDispose(ctx);
    return 1;
  }
  LLVMDisposeMessage(error);

  // Print diagnostic comments
  printf("; Test: test_builder_control_flow\n");
  printf(";\n");
  printf("; Control flow operations demonstrated:\n");
  printf(";   ret void, ret value\n");
  printf(";   br (unconditional)\n");
  printf(";   br (conditional)\n");
  printf(";   switch with 3 cases + default\n");
  printf(";   call\n");
  printf(";   unreachable\n");
  printf(";\n");
  printf("; Branch analysis:\n");
  printf(";   unconditional br is conditional: %s\n",
         LLVMIsConditional(br_inst) ? "yes" : "no");
  printf(";   conditional br is conditional: %s\n",
         LLVMIsConditional(cond_br) ? "yes" : "no");
  printf(";   unconditional br num successors: %u\n",
         LLVMGetNumSuccessors(br_inst));
  printf(";   conditional br num successors: %u\n",
         LLVMGetNumSuccessors(cond_br));
  printf(";\n");
  printf("; Current insert block: %s\n", LLVMGetBasicBlockName(current_block));
  printf("\n");

  // Print module IR
  char *ir = LLVMPrintModuleToString(mod);
  printf("%s", ir);
  LLVMDisposeMessage(ir);

  // Cleanup
  LLVMDisposeModule(mod);
  LLVMContextDispose(ctx);

  return 0;
}
